third_party/amd/lib/TritonAMDGPUToLLVM/LoadStoreOpToLLVM.cpp

```diff
@@ -511,7 +511,41 @@ struct DirectToLdsLoadConversionBase : public LoadStoreConversionBase {
         LLVM::getSharedMemoryObjectFromStruct(loc, llDst, resElemTy, rewriter);
     auto affineOffset = smemObj.getShmemOffset(loc, rewriter, dstTy);
     auto maskSpanAffineOffset = SharedMemoryObject::getMaskSpanOffsets(dstTy);
-    auto [laneId, warpId] = getLaneAndWarpId(rewriter, loc);
+
+    Value laneId, warpId;
+    if (ISAFamily::CDNA3 == isaFamily || ISAFamily::CDNA4 == isaFamily) {
+      // On GFX9, there is no dedicated hardware instruction to read `wave_id`.
+      // The value is instead computed from `workitem.id.x`. Per the GFX9 ABI,
+      // `workitem.id.x` is initialized in a vector register, and vector
+      // instructions are generated for IR operations that depend on `wave_id`.
+      //
+      // A `v_readfirstlane` instruction is inserted at the end of these vector
+      // sequences to transfer the value from a vector register to a scalar
+      // register, initializing `$m0`.
+
+      // When this sequence occurs inside a loop, the MachineLICM pass does not
+      // hoist it because `v_readfirstlane` is convergent. Since both
+      // `workitem.id.x` and `wave_id` are constant at runtime, their
+      // computation can be safely hoisted to the function entry block.
+      auto insertPt = rewriter.saveInsertionPoint();
+      Operation *parentOp = insertPt.getBlock()->getParentOp();
+      while (!isa<LLVM::LLVMFuncOp>(parentOp)) {
+        parentOp = parentOp->getParentOp();
+      }
+
+      assert(parentOp);
+      auto funcOp = dyn_cast<LLVM::LLVMFuncOp>(parentOp);
```

Comment:
Nit: You can just do `auto funcOp = cast<LLVM::LLVMFuncOp>(parent)` here which internally `assert` so you don't need to have L536.